Rollbar/RollbarPayloadTruncator.m

```diff
@@ -250,36 +250,40 @@ +(NSString*)truncateString:(NSString*)inputString
     
     unsigned long currentStringEncoodingBytes =
         [RollbarPayloadTruncator measureTotalEncodingBytes:inputString];
-    if (currentStringEncoodingBytes <= totalBytesLimit)
-    {
+    
+    if (currentStringEncoodingBytes <= totalBytesLimit) {
         // no need to truncate:
         return inputString;
     }
     
     NSString *ellipsis = @"...";
-    unsigned long totalEllipsisEncodingBytes =
+    
+    const unsigned long totalEllipsisEncodingBytes =
         [RollbarPayloadTruncator measureTotalEncodingBytes:ellipsis];
-    unsigned long bytesToRemove =
-        currentStringEncoodingBytes - totalBytesLimit + totalEllipsisEncodingBytes;
-
-    NSMutableString *result =
-        [NSMutableString stringWithString:
-         [inputString substringToIndex:inputString.length - bytesToRemove]
-         ];
-    [result appendString:ellipsis];
-    currentStringEncoodingBytes = [RollbarPayloadTruncator measureTotalEncodingBytes:result];
-    while (totalBytesLimit < currentStringEncoodingBytes) {
-        
-        bytesToRemove =
-            currentStringEncoodingBytes - totalBytesLimit + totalEllipsisEncodingBytes;
-        
-        [result deleteCharactersInRange:NSMakeRange(result.length - bytesToRemove, bytesToRemove)];
-        [result appendString:ellipsis];
-        
-        currentStringEncoodingBytes = [RollbarPayloadTruncator measureTotalEncodingBytes:result];
+    
+    if (totalEllipsisEncodingBytes >= totalBytesLimit) {
+        // we have to have at least the ellipsis as a reasult of a string truncation:
+        return ellipsis;
     }
     
-    return result;
+    NSMutableString *bestFit = [NSMutableString string];
+    NSMutableString *result = [NSMutableString string];
+    [inputString enumerateSubstringsInRange:NSMakeRange(0, inputString.length)
+                                    options:NSStringEnumerationByComposedCharacterSequences | NSStringEnumerationLocalized
+                                 usingBlock:^(NSString *letter, NSRange letterRange, NSRange enclosingLetterRange, BOOL *stop2) {
```

Comment:
This looks extremely slow. I'm not sure I even need to profile it to say that. Things that are slow:

* Building a string by appending single characters (we are doing this twice here)
* Measuring the size of a string in a given encoding (we are doing this for every character)

All we care about is getting the strings down to something smaller so that we can hopefully shrink the payload a little bit. This is way too much work for something that does not need to be that precise.

Reply:
agree - it is slow but most accurate so far and probably having lowest risk factor of possible crashes. so, let me try one more time going the way we had it before: less accurate but better performance while limiting risks of possible crashes.